File: Examples/GroupJson/src/cpp/Group.cpp
#include "Group.hpp"
#include <iostream>
#include <stdlib.h>
#include <stdio.h>
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <json/json.h>

using namespace std;

/**
 * Copyright 2015 Tim Lindquist,
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * 
 * Purpose: demonstrate json serialization and deserialization in C++
 * Ser321 Foundations of Distributed Applications
 * see http://pooh.poly.asu.edu/Ser321
 * @author Tim Lindquist Tim.Lindquist@asu.edu
 *         Software Engineering, CIDSE, IAFSE, ASU Poly
 * @version July 2015
 */
Group::Group(){
}

void Group::initGroupFromJsonFile(string jsonFileName){
   ifstream jsonFile(jsonFileName.c_str());
   string line;
   cout << "The content of the Json file as a string: " << endl;
   if(jsonFile.is_open()){
      while(getline(jsonFile,line)){
         cout << line << endl;
      }
   }else{
      cout << "Json file not opened properly" << endl;
   }
   jsonFile.close();
   Json::Reader reader;
   Json::Value root;
   std::ifstream json(jsonFileName.c_str(), std::ifstream::binary);
   bool parseSuccess = reader.parse(json,root,false);
   if(parseSuccess){
      //cout << "successful parse" << endl;
      Json::Value::Members mbr = root.getMemberNames();
      for(vector<string>::const_iterator i = mbr.begin(); i!= mbr.end(); i++){
         //cout << *i << " " << endl;
         Json::Value jsonUser = root[*i];
         string nameStr = "name";
         if(nameStr.compare(*i)==0){
            name = jsonUser.asString();
            //cout << "found name " << *i << " value: " << name << endl;
         }else{
            //cout << "found " << *i << endl;
            string userId = jsonUser["userId"].asString();
            string userPwd = jsonUser["userPwd"].asString();
            //cout << "pwd: " << userPwd << " user: " << userId << endl;
            User * user = new User(userId, userPwd);
            //user->print();
            users[*i] = *user;
         }
      }
   }
}

void Group::toJsonFile(string jsonFileName){
   Json::Value jsonLib;
   jsonLib["name"] = name;
   for(std::map<string,User>::iterator i = users.begin(); i!= users.end(); i++){
      string key = i->first;
      //cout << key << " " << endl;
      Json::Value jsonUser;
      User usr = users[key];
      jsonUser["userId"] = usr.userId;
      jsonUser["userPwd"] = usr.userPwd;
      jsonLib[key] = jsonUser;
   }
   Json::StyledStreamWriter ssw("  ");
   std::ofstream jsonOutFile(jsonFileName.c_str(), std::ofstream::binary);
   ssw.write(jsonOutFile, jsonLib);
}

Group::~Group() {
   //cout << "Group destructor.\n";
   users.clear();
}

void Group::add(User usr){
   users[usr.userId]=usr;
}

void Group::remove(User usr){
   users.erase(usr.userId);
}

User Group::get(string usrName){
   User ret = users[usrName];
   return ret;
}

std::vector<string> Group::getNames(){
   vector<string> myVec;
   for(map<string,User>::iterator it = users.begin();it!=users.end();++it){
      myVec.push_back(it->first);
      //cout << it->first << "\n";
   }
   return myVec;
}

void Group::printGroup(){
   cout << "Group " << name << " has users" << endl;
   for(map<string,User>::iterator it = users.begin();it!=users.end();++it){
      cout << "key " << it->first << " is ";
      (it->second).print();
   }
   cout << endl;

}

File: Examples/GroupJson/src/java/Group.java
package ser321.serialize;

import org.json.JSONString;
import org.json.JSONObject;
import org.json.JSONTokener;
import java.io.FileInputStream;
import java.util.Vector;
import java.util.Enumeration;
import java.io.Serializable;

/**
 * Copyright (c) 2015 Tim Lindquist,
 * Software Engineering,
 * Arizona State University at the Polytechnic campus
 * <p/>
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation version 2
 * of the License.
 * <p/>
 * This program is distributed in the hope that it will be useful,
 * but without any warranty or fitness for a particular purpose.
 * <p/>
 * Please review the GNU General Public License at:
 * http://www.gnu.org/licenses/gpl-2.0.html
 * see also: https://www.gnu.org/licenses/gpl-faq.html
 * so you are aware of the terms and your rights with regard to this software.
 * Or, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301,USA
 * <p/>
 * Purpose:
 * A class implementing the operations for group management.
 * <p/>
 *
 * Ser321 Principles of Distributed Software Systems
 * @see <a href="http://pooh.poly.asu.edu/Ser321">Ser321 Home Page</a>
 * @author Tim Lindquist (Tim.Lindquist@asu.edu) CIDSE - Software Engineering
 *                       Ira Fulton Schools of Engineering, ASU Polytechnic
 * @file    Group.java
 * @date    July, 2015
 * @license See above
 * @see ser321.serialize.User
 */
public class Group extends Object implements JSONString, Serializable {

   // Serial version UID is defined below. Its only needed if you want
   // to make changes to the class and still deserialize artifacts
   // generated from prior versions. Obtain this definition with:
   // serialver -classpath classes:lib/json.jar ser321.serialize.Group
   private static final long serialVersionUID = 6982142948226029575L;

   private String name; // the name of this group
   private Vector<User> users = new Vector<User>(); // the members of this group

   public Group() {
      this.name = "unknown";
   }

   public Group(String fileName){
      try{
         FileInputStream in = new FileInputStream(fileName);
         JSONObject obj = new JSONObject(new JSONTokener(in));
         String [] names = JSONObject.getNames(obj);
         System.out.print("names are: ");
         for(int j=0; j< names.length; j++){
            System.out.print(names[j]+", ");
         }
         System.out.println("");
         name = obj.getString("name");
         users = new Vector<User>();
         for (int i=0; i< names.length; i++){
            if(!names[i].equals("name")){
               User aUser = new User((JSONObject)obj.getJSONObject(names[i]));
               users.add(aUser);
            }
         }
         in.close();
      }catch (Exception ex) {
         System.out.println("Exception importing from json: "+ex.getMessage());
      }
   }

   public String toJSONString(){
      String ret;
      JSONObject obj = new JSONObject();
      obj.put("name",name);
      for (Enumeration<User> e = users.elements(); e.hasMoreElements();){
         User usr = (User)e.nextElement();
         obj.put(usr.getId(),usr.toJSONObject());
      }
      ret = obj.toString();
      //System.out.println("group tojsonstring returning string: "+ret);
      return ret;
   }

   public String getName(){
      return name;
   }

   public void setName(String aName){
      name = aName;
   }

  /**
   * Associate a new user with this authorization group.
   * Use addUserToGroup to allow a new user access to group resources.
   * @param user Is a String specifying userId to add
   * @param pwd Is a String specifying password for user.
   */
   public void addUserToGroup(String id, String pwd) {
      boolean found = false;
      for (int i = 0; i<users.size(); i++) {
         if ((users.elementAt(i)).getId().equals(id))
            found = true;
      }
      if (!found)
         users.addElement(new User(id,pwd));
   }

  /**
   * Get the name strings of all users associated in this authorization group.
   * Use getName to retrieve the name string property for the group.
   * @return The vector of userId strings.
   */
   public Vector<String> getUserNames() {
      Vector<String> ret = new Vector<String>();
      for (Enumeration e = users.elements() ; e.hasMoreElements() ;) {
         ret.addElement(((User)e.nextElement()).getId());
      }
      return ret;
   }

  /**
   * Determine whether a user is in the group.
   * Use isMember to authenticate a user password combination for this group
   * membership.
   * @param id Is a String specifying userId that is not already in the group
   * @param pwd Is a String specifying a password for id.
   * @return true if user and pwd are authorized;
   * otherwise return false.
   */
   public boolean isMember(String id, String pwd) {
      boolean match = false;
      for (Enumeration e = users.elements() ; e.hasMoreElements() ;) {
         if(((User)e.nextElement()).check(id,pwd)){
            match = true;
            break;
         }
      }
      return match;
   }

  /**
   * Print the group to standard out
   */
   public void printGroup(){
      System.out.print("Group: "+name+" has users: ");
      for (Enumeration e = users.elements() ; e.hasMoreElements() ;) {
         System.out.print(((User)e.nextElement()).getId()+", ");
      }
      System.out.println();
   }

}
